### PR TITLE
Fix S4487 FP: ??= operator reads value

### LINKED ISSUE #6653: S4487 FP with ??= operator.
### Description

I get an S4487 false positive with the ??= operator, the equivalent code using `if` does not trigger the rule.

### Repro steps
```
public class Class1
{
    /// <summary>
    /// Our delay to debounce when there are lots of rapid changes to the same file.
    /// </summary>
    private static readonly TimeSpan DebounceDelay = TimeSpan.FromSeconds(1);

    private readonly object _lock = new();

    // This line trigger the rule
    private Task? _checkerTask;

    private bool _dirty;

    public void TriggerBackgroundChecks()
    {
        lock (_lock)
        {
            _dirty = true;
            // start a new checker task when there's none running.
            _checkerTask ??= BackgroundCheckLoopAsync();
        }
    }

    /* This code with if does not trigger the rule. */
    // public void TriggerBackgroundChecksWithIf()
    // {
    //     lock (_lock)
    //     {
    //         _dirty = true;
    //         if (_checkerTask == null)
    //             _checkerTask = BackgroundCheckLoopAsync();
    //     }
    // }

    private async Task BackgroundCheckLoopAsync()
    {
        while (true)
        {
            lock (_lock)
            {
                if (!_dirty)
                {
                    // unregister the checker task and exit.
                    _checkerTask = null;
                    return;
                }

                _dirty = false;
            }

            // Perform some checks...
            await PerformExpensiveChecksAsync();

            // wait for the delay before we run the next check.
            await Task.Delay(DebounceDelay);
        }
    }

    private ValueTask PerformExpensiveChecksAsync() => ValueTask.CompletedTask;
}

```
### Expected behavior

As the variable is semantically important in this case (I only want a single Checker Task to run at the same time), removing the member as suggested by the rule makes no sense. So I expect the rule to not trigger in this example.

### Actual behavior

The rule triggers on the variable _checkerTask, and claims it's "unread", and I should remove it.

### Known workarounds

A workaround is to use the commented TriggerBackgroundChecksWithIf() method, which is semantically valid. But my IDE (JetBrains Rider) then suggests that I simplify the `if` statement to the equivalent `??=` assignment.

### Related information

* C#/VB.NET Plugins version: SonarLint für Rider 7.3.0.59206
* JetBrains Rider: 2022.3.1 Build #RD-223.8214.53
* MSBuild / dotnet version: .NET Sdk 7.0.100
* Operating System: Windows 10

### PR DESCRIPTION
**Pull request by antonioaversa**
Fixes #6653

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs [hunk 1]
```diff
@@ -82,3 +82,65 @@ private class MyServiceTransient : IMyService
...
+        public string Property10 => this[Field10 ??= "1"];
+        public string Property11 => this.Field11 ?? "1";
+        public string Property12 => Field12[0] ??= "1";
+        public string Property13 => Field13_1 ??= (Field13_2 += "1");
+        public string Property14 => Field14_1 ??= (Field14_2 = "1");
+        public string Property15 => Field15_1 ??= SomeMethod(Field15_2 = "1");
+
+        public UsingCoalescingAssignmentOperator(string sPar) { }
+
+        public string this[string sPar] => sPar + "1";
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs:139]: AFAICT, this is just scaffolding to enable the testcases above, right?

Let's just return null so we don't need to think about what `sPar` and `+"1"` is testring
```suggestion
        public string this[string sPar] => null;
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs [hunk 2]
```diff
@@ -82,3 +82,65 @@ private class MyServiceTransient : IMyService
...
+
+        public static UsingCoalescingAssignmentOperator StaticProperty1() => new UsingCoalescingAssignmentOperator(StaticField1 ??= "1");
+
+        private string SomeMethod(string sPar) => sPar + "1";
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs:143]: Same as above
```suggestion
        private string SomeMethod(string sPar) => null;
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs [hunk 3]
```diff
@@ -82,3 +82,65 @@ private class MyServiceTransient : IMyService
...
+        public string Property6 => Field6_1 ??= Field6_2 ??= Field6_3;
+        public string Property7 => Field7_1 ??= (Field7_2 ?? Field7_3);
+        public string Property8 => "1" + (Field8 ??= "1");
+        public string Property9 => SomeMethod(Field9 ??= "1");
+        public string Property10 => this[Field10 ??= "1"];
+        public string Property11 => this.Field11 ?? "1";
+        public string Property12 => Field12[0] ??= "1";
+        public string Property13 => Field13_1 ??= (Field13_2 += "1");
+        public string Property14 => Field14_1 ??= (Field14_2 = "1");
+        public string Property15 => Field15_1 ??= SomeMethod(Field15_2 = "1");
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs:135]: The fact that these are properties is not important here, right? Let's make it a method and assign all to "string value;". Something like
```
value = Field0_3;
value = Field ??= "1";
...
```

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 4]
```diff
@@ -259,6 +259,9 @@ node.Parent switch
                 MemberAccessExpressionSyntax memberAccess => node == memberAccess.Name ? ParentAccessType(memberAccess) : SymbolAccess.Read,
                 // _?.node : node?._
                 MemberBindingExpressionSyntax memberBinding => node == memberBinding.Name ? ParentAccessType(memberBinding) : SymbolAccess.Read,
+                // node ??= _ : _ ??= node
+                AssignmentExpressionSyntax assignmentExpression when assignmentExpression.IsKind(SyntaxKindEx.CoalesceAssignmentExpression) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:263]: Everything is expression and syntax around here :)
```suggestion
                AssignmentExpressionSyntax assignment when assignment.IsKind(SyntaxKindEx.CoalesceAssignmentExpression) =>
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs [hunk 5]
```diff
@@ -82,3 +82,65 @@ private class MyServiceTransient : IMyService
...
+// https://github.com/SonarSource/sonar-dotnet/issues/6653
+namespace Repro6653
+{
+    public class UsingCoalescingAssignmentOperator
+    {
+        private static string StaticField1; // Compliant, read and written by ??= in SomeStaticMethod
+
+        private string Field0_1;            // Noncompliant, never read nor written
+        private string Field0_2 = "1";      // Noncompliant, written by initializer but not read
+        private string Field0_3 = "1";      // Compliant, read by Property0 and written by initializer
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs:95]: I would remove all the `// Compliant` comments for maintainability reasons. This class is focused on coalescing and quick Ctrl+F shows the usage and reason pretty clearly.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Nice, just some cleanup in UTs

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs (hunk 3 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnusedPrivateMember.CSharp8.cs:135]: Indeed! Fixed.

As the shadok used to say: "pourquoi faire simple quand on peut faire compliqué" :-)

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**